core/services/mongodb-remote/src/main/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/DataSynchronizer.java

```diff
@@ -969,435 +895,571 @@ private void syncLocalToRemote() {
+                        return localChangeEvent.getOperationType().toString();
+                      }
+                    };
                 }
-                break;
+              } else {
+                nextVersion = null;
               }
 
-              case DELETE: {
-                nextVersion = null;
-                final RemoteDeleteResult result;
-                // a. Delete the document in the remote database using a query for the _id and the
-                //    version.
-                try {
-                  result = remoteColl.deleteOne(localVersionInfo.getFilter());
-                } catch (final StitchServiceException ex) {
-                  // b. If an error happens, report an error to the error listener.
-                  emitError(
-                      docConfig,
-                      String.format(
-                          Locale.US,
-                          "t='%d': syncLocalToRemote ns=%s documentId=%s exception "
-                              + " deleting: %s",
-                          logicalT,
-                          nsConfig.getNamespace(),
-                          docConfig.getDocumentId(),
-                          ex),
-                      ex
-                  );
-                  continue;
+              if (logger.isDebugEnabled()) {
+                logger.debug(String.format(
+                    Locale.US,
+                    "t='%d': syncLocalToRemote ns=%s documentId=%s conflict=%s",
+                    logicalT,
+                    nsConfig.getNamespace(),
+                    docConfig.getDocumentId(),
+                    action == SyncAction.CONFLICT));
+              }
+
+              if (action != SyncAction.CONFLICT) {
+                // iv. If no conflict has occurred, move on to the remote to local sync routine.
+
+                // since we strip version information from documents before setting pending writes,
+                // we don't have to worry about a stale document version in the event here.
+                final ChangeEvent<BsonDocument> committedEvent =
+                    docConfig.getLastUncommittedChangeEvent();
+                emitEvent(docConfig.getDocumentId(), new ChangeEvent<>(
+                    committedEvent.getId(),
+                    committedEvent.getOperationType(),
+                    committedEvent.getFullDocument(),
+                    committedEvent.getNamespace(),
+                    committedEvent.getDocumentKey(),
+                    committedEvent.getUpdateDescription(),
+                    false));
+
+                setPendingWritesComplete = true;
+                if (committedEvent.getOperationType() != OperationType.DELETE) {
+                  localSyncWriteModelContainer.addConfigWrite(
+                      new ReplaceOneModel<>(CoreDocumentSynchronizationConfig.getDocFilter(
+                          nsConfig.getNamespace(), docConfig.getDocumentId()),
+                          docConfig));
```

Comment:
[q] it's a little unclear that the `docConfig` gets mutated later before the write is committed. Why don't we just call `docConfig.setPendingWritesComplete(nextVersion)` before the `addConfigWrite` call so it's clear that's what we're trying to convey with the local config write?

Reply:
because we have a read lock at this point and setPendingWritesComplete requires a write lock

Reply:
Alright, in that case can you add a quick comment then explaining that the doc config will have its pending writes set as complete by the time the write model commits its writes? I could see a future change accidentally messing up this logic.